File: bank-service/src/main/java/rs/edu/raf/service/impl/CardServiceImpl.java
package rs.edu.raf.service.impl;

import lombok.AllArgsConstructor;
import org.springframework.stereotype.Service;
import rs.edu.raf.dto.CardDTO;
import rs.edu.raf.dto.CreateCardDto;
import rs.edu.raf.exceptions.CardDeactivatedException;
import rs.edu.raf.exceptions.CardNotFoundException;
import rs.edu.raf.model.card.Card;
import rs.edu.raf.mapper.CardMapper;
import rs.edu.raf.model.card.CardStatus;
import rs.edu.raf.repository.CardRepository;
import rs.edu.raf.service.CardService;

import java.util.List;

@Service
@AllArgsConstructor
public class CardServiceImpl implements CardService {
    private final CardRepository cardRepository;
    private final CardMapper cardMapper;

    @Override
    public CardDTO createCard(CreateCardDto createCardDto){
        return cardMapper.cardToCardDto(cardRepository.save(cardMapper.createCartDtoToCart(createCardDto)));
    }

    @Override
    public void deleteCard(String cardNumber){
        Card card = cardRepository.findByNumber(cardNumber)
                .orElseThrow(()->new CardNotFoundException(cardNumber));
        cardRepository.delete(card);
    }

    @Override
    public void blockCard(String cardNumber){
        Card card = cardRepository.findByNumber(cardNumber)
                .orElseThrow(()->new CardNotFoundException(cardNumber));

        if(card.getCardStatus().equals(CardStatus.DEACTIVATED))
            throw new CardDeactivatedException(cardNumber);

        card.setCardStatus(CardStatus.BLOCKED);
        cardRepository.save(card);
    }

    @Override
    public void unblockCard(String cardNumber) {
        Card card = cardRepository.findByNumber(cardNumber)
                .orElseThrow(()->new CardNotFoundException(cardNumber));

        if(card.getCardStatus().equals(CardStatus.DEACTIVATED))
            throw new CardDeactivatedException(cardNumber);

        card.setCardStatus(CardStatus.ACTIVE);
        cardRepository.save(card);
    }

    @Override
    public void deactivateCard(String cardNumber){
        Card card = cardRepository.findByNumber(cardNumber)
                .orElseThrow(()->new CardNotFoundException(cardNumber));

        card.setCardStatus(CardStatus.DEACTIVATED);
        cardRepository.save(card);
    }

    @Override
    public List<CardDTO> getAllCards(){
        return cardRepository.findAll().stream().map(cardMapper::cardToCardDto).toList();
    }

    @Override
    public List<CardDTO> getAllCardsForUser(Long userId){
        return cardRepository.findByBankAccount_Owner(userId).stream().map(cardMapper::cardToCardDto).toList();
    }

}

File: bank-service-kotlin/src/main/kotlin/org/example/bankservicekotlin/service/impl/CardServiceImpl.kt
package org.example.bankservicekotlin.service.impl

import org.example.bankservicekotlin.dto.CardDTO
import org.example.bankservicekotlin.dto.CreateCardDTO
import org.example.bankservicekotlin.exception.CardDeactivatedException
import org.example.bankservicekotlin.exception.CardNotFoundException
import org.example.bankservicekotlin.mapper.CardMapper
import org.example.bankservicekotlin.model.card.Card
import org.example.bankservicekotlin.model.card.CardStatus
import org.example.bankservicekotlin.repository.CardRepository
import org.example.bankservicekotlin.service.CardService
import org.springframework.stereotype.Service

@Service
class CardServiceImpl(
    private val cardRepository: CardRepository,
    private val cardMapper: CardMapper
): CardService {
    override fun createCard(createCardDto: CreateCardDTO): CardDTO {
        return cardMapper.cardToCardDto(cardRepository.save(cardMapper.createCartDtoToCart(createCardDto)))
    }

    override fun deleteCard(cardNumber: String) {
        val card: Card = cardRepository.findByNumber(cardNumber)
            .orElseThrow { CardNotFoundException(cardNumber) }
        cardRepository.delete(card)
    }

    override fun blockCard(cardNumber: String) {
        val card: Card = cardRepository.findByNumber(cardNumber)
            .orElseThrow { CardNotFoundException(cardNumber) }

        if (card.cardStatus == CardStatus.DEACTIVATED) throw CardDeactivatedException(cardNumber)

        card.cardStatus = CardStatus.BLOCKED
        cardRepository.save(card)
    }

    override fun unblockCard(cardNumber: String) {
        val card: Card = cardRepository.findByNumber(cardNumber)
            .orElseThrow { CardNotFoundException(cardNumber) }

        if (card.cardStatus == CardStatus.DEACTIVATED) throw CardDeactivatedException(cardNumber)

        card.cardStatus = CardStatus.ACTIVE
        cardRepository.save(card)
    }

    override fun deactivateCard(cardNumber: String) {
        val card: Card = cardRepository.findByNumber(cardNumber)
            .orElseThrow { CardNotFoundException(cardNumber) }

        card.cardStatus = CardStatus.DEACTIVATED
        cardRepository.save(card)
    }

    override fun getAllCards(): List<CardDTO> {
        return cardRepository.findAll().stream().map(cardMapper::cardToCardDto).toList()
    }

    override fun getAllCardsForUser(userId: Long): List<CardDTO> {
        return cardRepository.findByBankAccount_Owner(userId).stream().map(cardMapper::cardToCardDto).toList()
    }
}